# Data Analysis & Prediction Model 
***

### Welcome!
In this project, we will be analysing a `Sales Dataset` for a ***Fictional Enterprise*** for better understanding of its sales in different categories (*Electronics, Clothing, Food etc.*) as well as to understand the sales of different brands. We will be using $ Python $ and its libraries which include the following :-
* _Pandas_
* _Numpy_
* _Matplotlib_

*Lets us begin our journey!*

## Importing The Dataset
***
Our first step will be to `import` the necessary ***Libraries*** and the ***Dataset*** to our *working environment*. The following lines of code will illustrate it.

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Sales_Data = pd.read_csv(r"C:\Users\Rudransh\Desktop\Enterprise Sales.csv")

Sales_Data.head()

Date     Day               Product Name  Quantity  Price
0  01-01-2024  Monday  Ferrero Rocher Chocolates        12    672
1  01-01-2024  Monday               Axe Perfumes         2   1262
2  01-01-2024  Monday      Tommy Hilfiger Shirts         1   2354
3  01-01-2024  Monday         Urban Ladder Table         2  20195
4  01-01-2024  Monday              Peppy's Chips        16    336

As we can see above, we have successfully imported libraries and the dataset and managed to display the first 5 rows of the dataset using the **Pandas DataFrame** `head()` function. The dataset is stored in the variable `Sales_Data` which we will be using to represent the entire dataset. To get the general overview of the dataset, we use the **DataFrame** `info()` function.

In [5]:
Sales_Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 5 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   Date          1000000 non-null  object
 1   Day           1000000 non-null  object
 2   Product Name  1000000 non-null  object
 3   Quantity      1000000 non-null  int64 
 4   Price         1000000 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 38.1+ MB


From the above information, we can conclude that there are a total of **1,000,000 Rows** in the dataset none of which are **null** values. There are 5 columns and 2 datatypes which are:
1. ***Date, Day, Product Name*** with datatype `object`
2. ***Quantity, Price*** with datatype `int64`

We will be filtering and categorising our data using the above columns.

## Categorising the Dataset
***
### 1. Categorising Data by Month
It is important to filter the dataset on **Monthly** basis to get a better idea of the Sales Pattern of the dataset.

To filter the data for every month, we can use the `loc[]` function, and input the range of dates for each month to get the monthly datasets. But before we can filter the data by **Date**, we need to set the **Date** column as the index of the dataset. This can be done by using the `set_index()` function.

In [8]:
Date_Sort = Sales_Data.set_index("Date")
Date_Sort

Day               Product Name  Quantity  Price
Date                                                           
01-01-2024   Monday  Ferrero Rocher Chocolates        12    672
01-01-2024   Monday               Axe Perfumes         2   1262
01-01-2024   Monday      Tommy Hilfiger Shirts         1   2354
01-01-2024   Monday         Urban Ladder Table         2  20195
01-01-2024   Monday              Peppy's Chips        16    336
...             ...                        ...       ...    ...
31-12-2024  Tuesday             H&M Kids Pants         1   2406
31-12-2024  Tuesday            Swayam Bedsheet         4    250
31-12-2024  Tuesday             GM Switchboard         1   3500
31-12-2024  Tuesday               Gap T-shirts         2   1174
31-12-2024  Tuesday           Onion Vegetables         6    372

[1000000 rows x 4 columns]

Now that the the index is set to **Date** column, we can filter the data for each month. We have stored this dataset in a new variable ***Date_Sort*** so as to not alter the original dataset.

In [10]:
Jan_Sales = Date_Sort.loc["01-01-2024" : "31-01-2024"]    # Sales data for January
Feb_Sales = Date_Sort.loc["01-02-2024" : "29-02-2024"]    # Sales data for February
Mar_Sales = Date_Sort.loc["01-03-2024" : "31-03-2024"]    # Sales data for March
Apr_Sales = Date_Sort.loc["01-04-2024" : "30-04-2024"]    # Sales data for April
May_Sales = Date_Sort.loc["01-05-2024" : "31-05-2024"]    # Sales data for May
Jun_Sales = Date_Sort.loc["01-06-2024" : "30-06-2024"]    # Sales data for June
Jul_Sales = Date_Sort.loc["01-07-2024" : "31-07-2024"]    # Sales data for July
Aug_Sales = Date_Sort.loc["01-08-2024" : "31-08-2024"]    # Sales data for August
Sep_Sales = Date_Sort.loc["01-09-2024" : "30-09-2024"]    # Sales data for September
Oct_Sales = Date_Sort.loc["01-10-2024" : "31-10-2024"]    # Sales data for October
Nov_Sales = Date_Sort.loc["01-11-2024" : "30-11-2024"]    # Sales data for November
Dec_Sales = Date_Sort.loc["01-12-2024" : "31-12-2024"]    # Sales data for December

We have stored the data for each month in the variables named after their respective months. Hence we can access the data for any month by calling its variable. For example, sales data for the month of **November** is shown below:

In [12]:
Nov_Sales

Day               Product Name  Quantity   Price
Date                                                             
01-11-2024    Friday              Cinthol Soaps        12    1728
01-11-2024    Friday              Peppy's Chips         1      32
01-11-2024    Friday                   Sony TVs         1   56615
01-11-2024    Friday              Bagrry's Oats        12     492
01-11-2024    Friday                 Gap Shirts         4   18916
...              ...                        ...       ...     ...
30-11-2024  Saturday    Cartier Women's Watches         1  187876
30-11-2024  Saturday  Cremica Fish Frozen Foods         3     798
30-11-2024  Saturday               Kissan Juice         5     440
30-11-2024  Saturday        Broccoli Vegetables        14    1260
30-11-2024  Saturday            Godrej Cupboard         4   21452

[79048 rows x 4 columns]

Hence from the above we can see the sales data for the month of November.

### 2. Categorising Data by Products

We will now categorise our data by **Product Categories** which will be very useful for analysing data more thouroughly and get more insights on Sales patterns. 

We will categorise the products in **7** different categories :

1. **Food**
2. **Electronics**
3. **Clothing**
4. **Home**
5. **Kids**
6. **Hygiene**
7. **Luxury**

To categorise our data in each of these categories, we will use keywords corresponding to each of their respective categories ( **E.g.** TVs , Speakers etc. in *Electronics* ; Shirts , Jeans etc. in *Clothing* and so on. ) and we can also use the brand name that can correspond to a category (**E.g.** Sony , Samsung , LG etc sells products in *Electronics* ; Kellogg's , Britannia etc sells products in *Food* and so on. ). These keywords will be put into the `str.contains()` function where it will search the entire dataset for rows containing the specified keywords and give us a new dataset based on each category.

After the data for each category has been filtered out, we will store the dataset in a new file using the `to_csv()` function so as to maintain records on each category of the products. Let us begin in order.

#### 1. Food

In [16]:
Food_Keywords = "Juice|Pulses|Cornflakes|Biscuits|Wheat|Rice|Chocolates|Cheese|Butter|Ice Cream|Popcorn|Oats|Vegetables|Fruits|Noodles|Frozen Food|Pasta|Cooking Oil|Chips|Soft Drinks|Cakes|Bread"
   
Food = Sales_Data[(Sales_Data["Product Name"].str.contains(Food_Keywords))]

Food

Date      Day               Product Name  Quantity  Price
0       01-01-2024   Monday  Ferrero Rocher Chocolates        12    672
4       01-01-2024   Monday              Peppy's Chips        16    336
6       01-01-2024   Monday             Pringles Chips        12    324
7       01-01-2024   Monday          Potato Vegetables        15    975
11      01-01-2024   Monday              Fortune Wheat        10    670
...            ...      ...                        ...       ...    ...
999979  31-12-2024  Tuesday           Parle-G Biscuits        24    240
999983  31-12-2024  Tuesday                Amul Cheese        17   2329
999987  31-12-2024  Tuesday            Britannia Cakes         5    855
999994  31-12-2024  Tuesday            Unibic Biscuits        24    456
999999  31-12-2024  Tuesday           Onion Vegetables         6    372

[400000 rows x 5 columns]

Thus we can see 400000 entries in **Food** Category. We now find the remaining categories.

#### 2. Electronics

In [19]:
Electronics_Keywords = "TVs|Laptops|Mobiles|MacBook|Speakers|Air Filters|Vaccum|Lights|Microwaves|ACs|OLED|Refrigerator|Washing Machine|QLED"
   
Electronics = Sales_Data[(Sales_Data["Product Name"].str.contains(Electronics_Keywords))]

Electronics

Date      Day             Product Name  Quantity   Price
9       01-01-2024   Monday    Honeywell Air Filters         1   36416
13      01-01-2024   Monday           Lenovo Laptops         1   61724
32      01-01-2024   Monday  Bang & Olufsen Speakers         2  190122
60      01-01-2024   Monday             Oppo Mobiles         1   21214
69      01-01-2024   Monday        Dyson Air Filters         1   38792
...            ...      ...                      ...       ...     ...
999926  31-12-2024  Tuesday  Apple iPhone 15 Mobiles         2  248282
999976  31-12-2024  Tuesday              Samsung ACs         1   42676
999978  31-12-2024  Tuesday     Panasonic Microwaves         1    9552
999985  31-12-2024  Tuesday      Sennheiser Speakers         1   12546
999986  31-12-2024  Tuesday  Bang & Olufsen Speakers         1   65696

[55000 rows x 5 columns]

#### 3. Clothing

In [21]:
Clothing_Keywords = "Shirts|Sweaters|Hoodies|Jackets|Shoes|Shorts|Cargo|Trousers|Sportswear|Sarees|Women's Wear|Innerwear|Underwear|Bras|Lingerie|Denim|Jeans|T-shirts|Belts"

Clothing_Exclude = "Balmain Paris|Christian Dior|Hermes|Louis Vuitton|Prada|Versace|Kids"

Clothing = Sales_Data[(Sales_Data["Product Name"].str.contains(Clothing_Keywords))]

Clothing = Clothing[(Clothing["Product Name"].str.contains(Clothing_Exclude) == False)] #To remove unwanted entries from other Categories

Clothing

Date      Day           Product Name  Quantity  Price
2       01-01-2024   Monday  Tommy Hilfiger Shirts         1   2354
10      01-01-2024   Monday       Jockey Underwear         2   3360
21      01-01-2024   Monday          Zara Sweaters         1    483
25      01-01-2024   Monday           Levi's Cargo         1   3104
27      01-01-2024   Monday           Zara Hoodies         1   5698
...            ...      ...                    ...       ...    ...
999990  31-12-2024  Tuesday             Puma Shoes         1   6567
999991  31-12-2024  Tuesday   Decathlon Sportswear         1    799
999992  31-12-2024  Tuesday         Wrangler Cargo         1   1087
999993  31-12-2024  Tuesday            Zara Shorts         2   3332
999998  31-12-2024  Tuesday           Gap T-shirts         2   1174

[150000 rows x 5 columns]

In the above codes we had some entries from other categories being included with current category, Therefore we gave a `False` Condition to the keywords that should not be included.

#### 4. Home

In [24]:
Home_Keywords = "Switchboard|Decor|Fan|Bedsheet|Dining|Curtain|Chair|Cupboard|Sofa|Bed|Mirror|Shelf|Table|Pillow"

Home_Exclude = "Fanta|Biscuits"

Home = Sales_Data[(Sales_Data["Product Name"].str.contains(Home_Keywords))]

Home = Home[(Home["Product Name"].str.contains(Home_Exclude) == False)] #To remove unwanted entries from other Categories

Home

Date      Day        Product Name  Quantity  Price
3       01-01-2024   Monday  Urban Ladder Table         2  20195
19      01-01-2024   Monday      Corelle Dining         1  12325
26      01-01-2024   Monday     Pepperfry Decor         4    200
57      01-01-2024   Monday           Bajaj Fan         6   1750
63      01-01-2024   Monday     Spacewood Table         1  23204
...            ...      ...                 ...       ...    ...
999960  31-12-2024  Tuesday        Duroflex Bed         1  34571
999971  31-12-2024  Tuesday  Anchor Switchboard         4    919
999977  31-12-2024  Tuesday     La Opala Dining         7  15886
999996  31-12-2024  Tuesday     Swayam Bedsheet         4    250
999997  31-12-2024  Tuesday      GM Switchboard         1   3500

[150000 rows x 5 columns]

#### 5. Kids

In [28]:
Kids_Keywords = "Kids|Baby|Action Figure|Board Game|Building Blocks|Car|Doll|Lego|Teddy|Stuffed Animal|RC|Puzzle"

Kids_Exclude = "Cargo|Oral Care|Skin Care|Cartier|Vegetables"

Kids = Sales_Data[(Sales_Data["Product Name"].str.contains(Kids_Keywords))]

Kids = Kids[(Kids["Product Name"].str.contains(Kids_Exclude) == False)] #To remove unwanted entries from other Categories

Kids

Date      Day            Product Name  Quantity  Price
8       01-01-2024   Monday       Next Kids Sweater         5   2316
17      01-01-2024   Monday   Disney Stuffed Animal         4   4767
24      01-01-2024   Monday        Zara Kids Jacket         6   1694
35      01-01-2024   Monday   Hot Wheels Teddy Bear         4   2324
56      01-01-2024   Monday  Burt's Bees Baby Wipes         3    564
...            ...      ...                     ...       ...    ...
999967  31-12-2024  Tuesday     Bratz Action Figure         1   3372
999968  31-12-2024  Tuesday    LEGO Building Blocks         5   4573
999984  31-12-2024  Tuesday           Nerf Lego Set         2   3963
999988  31-12-2024  Tuesday   Johnson's Baby Powder        14    137
999995  31-12-2024  Tuesday          H&M Kids Pants         1   2406

[90000 rows x 5 columns]

#### 6. Hygiene

In [31]:
Hygiene_Keywords = "Perfumes|Soaps|Oral Care|Shampoo|Skin Care|Body Wash|Hygiene"

Hygiene_Exclude = "Baby"

Hygiene = Sales_Data[(Sales_Data["Product Name"].str.contains(Hygiene_Keywords))]

Hygiene = Hygiene[(Hygiene["Product Name"].str.contains(Hygiene_Exclude) == False)] #To remove unwanted entries from other Categories

Hygiene

Date      Day             Product Name  Quantity  Price
1       01-01-2024   Monday             Axe Perfumes         2   1262
5       01-01-2024   Monday            Cinthol Soaps        11   1067
14      01-01-2024   Monday      Pepsodent Oral Care         1    286
30      01-01-2024   Monday          Sunsilk Shampoo         4    768
31      01-01-2024   Monday   Whisper Female Hygiene         6   7776
...            ...      ...                      ...       ...    ...
999878  31-12-2024  Tuesday           Olay Skin Care         8   4936
999883  31-12-2024  Tuesday           Olay Skin Care         7   2765
999911  31-12-2024  Tuesday  Stayfree Female Hygiene         7  12992
999928  31-12-2024  Tuesday             Dove Shampoo         6   4494
999980  31-12-2024  Tuesday          Pantene Shampoo         2   1284

[150000 rows x 5 columns]

#### 7. Luxury

In [33]:
Luxury_Keywords = "Audemars Piguet|Balmain Paris|Bulgari|Cartier|Chanel|Christian Dior|Czapek|Gucci|Hermes|Louis Vuitton|Patek Philippe|Prada|Rado|Rolex|Richard Mille|TAG Heuer|Versace"

Luxury = Sales_Data[(Sales_Data["Product Name"].str.contains(Luxury_Keywords))]

Luxury

Date      Day                    Product Name  Quantity   Price
101     01-01-2024   Monday         Cartier Women's Watches         1  164941
673     01-01-2024   Monday                  Gucci Handbags         1   94444
932     01-01-2024   Monday            Christian Dior Coats         1  137020
1236    01-01-2024   Monday             Balmain Paris Coats         1   89422
1366    01-01-2024   Monday         TAG Heuer Men's Watches         1  195725
...            ...      ...                             ...       ...     ...
998537  31-12-2024  Tuesday         Christian Dior Handbags         1  234668
998647  31-12-2024  Tuesday                 Hermes Handbags         1   75261
998952  31-12-2024  Tuesday                 Hermes Handbags         2  309942
999251  31-12-2024  Tuesday  Christian Dior Women's Jackets         1   59158
999594  31-12-2024  Tuesday                 Chanel Handbags         1   57723

[5000 rows x 5 columns]

We have successfully managed to categorise our products into the 7 specified categories. We can now move onto ***Analysing*** the dataset.

## Analyzing The Dataset
***
### 1. General Analysis
***
#### 1.1. Yearly Sales
We get the overview of the total sales made by the company during the timeframe **01-01-2024** to **31-12-2024**, by selecting the *price* column and totalling all the elements in it with use of the function `sum()`.

In [38]:
Total_Sales = Sales_Data[["Price"]]
Total_Sales

Price
0         672
1        1262
2        2354
3       20195
4         336
...       ...
999995   2406
999996    250
999997   3500
999998   1174
999999    372

[1000000 rows x 1 columns]

In [98]:
print("Total sales made during the period of 1 year (2024-24) = Rs.", Sales_Data["Price"].sum())

Total sales made during the period of 1 year (2024-24) = Rs. 6995741459


Thus we see that the total sales made in 1 year is **Rs. 6,995,741,459** or **Rs. 699.5 Crores.**

#### 1.2. Yearly Products Sold
To see the the total amount of products sold in a year, we use the `sum()` function on the **Quantity** column.

In [41]:
Total_Products = Sales_Data[["Quantity"]]
Total_Products

Quantity
0             12
1              2
2              1
3              2
4             16
...          ...
999995         1
999996         4
999997         1
999998         2
999999         6

[1000000 rows x 1 columns]

In [43]:
print("Total Products sold during the period of 1 year (2024-25) =",Sales_Data["Quantity"].sum(),"Products")

Total Products sold during the period of 1 year (2024-25) = 6927170 Products


Therefore total products sold during the period is **6927170 Products**

#### 1.3. Monthly Sales
Now that we have Sales data for each month, we can now calculate the total sales for each month using `sum()` function.

In [102]:
Months = ["January" , "February" , "March", "April" , "May" , "June" , 
         "July" , "August" , "September" , "October" , "November" , "December"]

Sales_Per_Month = [Jan_Sales["Price"] , Feb_Sales["Price"] , Mar_Sales["Price"] , Apr_Sales["Price"] , May_Sales["Price"] , Jun_Sales["Price"] , 
                   Jul_Sales["Price"] , Aug_Sales["Price"] , Sep_Sales["Price"] , Oct_Sales["Price"] , Nov_Sales["Price"] , Dec_Sales["Price"]]

Total_Monthly_Sales = []

print("Monthly Sales Report : \n")

for i in range(0 , 12):
    Total_Monthly_Sales.append(Sales_Per_Month[i].sum())
    print(i+1,".",Months[i], "=" ,Sales_Per_Month[i].sum())

Monthly Sales Report : 

1 . January = 578420576
2 . February = 572246415
3 . March = 594353266
4 . April = 575563124
5 . May = 588770072
6 . June = 589022390
7 . July = 577530505
8 . August = 601679385
9 . September = 567235742
10 . October = 593027477
11 . November = 571079665
12 . December = 586812842


#### 1.4. Monthly Products Sold
Similarly, Let us find the total Products sold in each month

In [50]:
Products_Sold_Per_Month = [Jan_Sales["Quantity"] , Feb_Sales["Quantity"] , Mar_Sales["Quantity"] , Apr_Sales["Quantity"] ,
                           May_Sales["Quantity"] , Jun_Sales["Quantity"] , Jul_Sales["Quantity"] , Aug_Sales["Quantity"] , 
                           Sep_Sales["Quantity"] , Oct_Sales["Quantity"] , Nov_Sales["Quantity"] , Dec_Sales["Quantity"]]

Total_Monthly_Products_Sold = []

print("Monthly Products Sold : \n")

for i in range(0 , 12):
    Total_Monthly_Products_Sold.append(Products_Sold_Per_Month[i].sum())
    print(i+1,".",Months[i], "=" ,Products_Sold_Per_Month[i].sum())

Monthly Products Sold : 

1 . January = 584235
2 . February = 565717
3 . March = 613077
4 . April = 562401
5 . May = 594549
6 . June = 591847
7 . July = 583492
8 . August = 592907
9 . September = 546301
10 . October = 567877
11 . November = 554806
12 . December = 569961


We have successfully calculated the Monthly sales and Monthly Products Sold and added them to thier respective lists for easier access. It will also be convenient to visually represent the above data in the form of a **graph**. Since the Monthly sales and Monthly Products data are similar to each other, it is beneficial to use a **Line Chart** for its representation. We will use the **Matplotlib** library to visualise the data.

#### 1.5. Day-Wise Sales

Let us now find the Day-Wise sales data to help us understand the sales pattern of the specific days.

In [84]:
Days = ["Monday" , "Tuesday" , "Wednesday" , "Thursday" , "Friday" , "Saturday" , "Sunday"]

print("Day-Wise Sales pattern : " , "\n")

gggg = []

for i in range(0,7):
    print(i+1,'.', Days[i], '=', Sales_Data[(Sales_Data["Day"] == Days[i])]["Price"].sum())

Day-Wise Sales pattern :  

1 . Monday = 702355744
2 . Tuesday = 705954185
3 . Wednesday = 934237030
4 . Thursday = 985593303
5 . Friday = 960306789
6 . Saturday = 1332207542
7 . Sunday = 1375086866


#### 1.6. Day-Wise Products Sold

In [78]:
print("Day-Wise Products Sold : " , "\n")

for i in range(0,7):
    print(i+1,'.', Days[i], '=', Sales_Data[(Sales_Data["Day"] == Days[i])]["Quantity"].sum())

Day-Wise Products Sold :  

1 . Monday = 605803
2 . Tuesday = 614135
3 . Wednesday = 974517
4 . Thursday = 986311
5 . Friday = 983385
6 . Saturday = 1380339
7 . Sunday = 1382680
